pkg/server/plugin/keymanager/hashicorpvault/vault_client.go

```diff
@@ -0,0 +1,554 @@
+		clientTLSConfig.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return &clientCert, nil
+		}
+	}
+
+	return nil
+}
+
+// SetToken wraps vapi.Client.SetToken()
+func (c *Client) SetToken(v string) {
+	c.vaultClient.SetToken(v)
+}
+
+// Auth authenticates to vault server with TLS certificate method
+func (c *Client) Auth(path string, body map[string]any) (*vapi.Secret, error) {
+	c.vaultClient.ClearToken()
+	secret, err := c.vaultClient.Logical().Write(path, body)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "authentication failed %v: %v", path, err)
+	}
+
+	tokenID, err := secret.TokenID()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "authentication is successful, but could not get token: %v", err)
+	}
+	c.vaultClient.SetToken(tokenID)
+	return secret, nil
+}
+
+func (c *Client) LookupSelf(token string) (*vapi.Secret, error) {
+	if token == "" {
+		return nil, status.Error(codes.InvalidArgument, "token is empty")
+	}
+	c.SetToken(token)
+
+	secret, err := c.vaultClient.Logical().Read("auth/token/lookup-self")
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "token lookup failed: %v", err)
+	}
+
+	id, err := secret.TokenID()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get TokenID: %v", err)
+	}
+	renewable, err := secret.TokenIsRenewable()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to determine if token is renewable: %v", err)
+	}
+	ttl, err := secret.TokenTTL()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get token ttl: %v", err)
+	}
+	secret.Auth = &vapi.SecretAuth{
+		ClientToken:   id,
+		Renewable:     renewable,
+		LeaseDuration: int(ttl.Seconds()),
+		// don't care any parameters
+	}
+	return secret, nil
+}
+
+type TransitKeyType string
+
+const (
+	TransitKeyTypeRSA2048   TransitKeyType = "rsa-2048"
+	TransitKeyTypeRSA4096   TransitKeyType = "rsa-4096"
+	TransitKeyTypeECDSAP256 TransitKeyType = "ecdsa-p256"
+	TransitKeyTypeECDSAP384 TransitKeyType = "ecdsa-p384"
+)
+
+type TransitHashAlgorithm string
+
+const (
+	TransitHashAlgorithmSHA256 TransitHashAlgorithm = "sha2-256"
+	TransitHashAlgorithmSHA384 TransitHashAlgorithm = "sha2-384"
+	TransitHashAlgorithmSHA512 TransitHashAlgorithm = "sha2-512"
+	TransitHashAlgorithmNone   TransitHashAlgorithm = "none"
+)
+
+type TransitSignatureAlgorithm string
+
+const (
+	TransitSignatureSignatureAlgorithmPSS      TransitSignatureAlgorithm = "pss"
+	TransitSignatureSignatureAlgorithmPKCS1v15 TransitSignatureAlgorithm = "pkcs1v15"
+)
+
+// CreateKey creates a new key in the specified transit secret engine
+// See: https://developer.hashicorp.com/vault/api-docs/secret/transit#create-key
+func (c *Client) CreateKey(ctx context.Context, spireKeyID string, keyType TransitKeyType) error {
+	arguments := map[string]interface{}{
+		"type":       keyType,
+		"exportable": "false", // TODO: Maybe make this configurable
```

Comment:
I don't think that we would want to make SPIRE keys exportable.

Reply:
In that case I remove the TODO.